### PR TITLE
[SPARK-41233][SQL][PYTHON] Add `array_prepend` function

### PR DESCRIPTION
**Pull request by navinvishy**
### What changes were proposed in this pull request?
Adds a new array function array_prepend to catalyst.

### Why are the changes needed?
This adds a function that exists in many SQL implementations, specifically Snowflake: https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/api/snowflake.snowpark.functions.array_prepend.html

### Does this PR introduce _any_ user-facing change?
Yes.

### How was this patch tested?
Added unit tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 1]
```diff
@@ -1376,35 +1418,148 @@ case class ArrayContains(left: Expression, right: Expression)
...
 
   override def prettyName: String = "array_contains"
 
   override protected def withNewChildrenInternal(
-      newLeft: Expression, newRight: Expression): ArrayContains =
+      newLeft: Expression,
+      newRight: Expression): ArrayContains =
     copy(left = newLeft, right = newRight)
 }
+@ExpressionDescription(
```
Comment (navinvishy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1444]: Relevant diff here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 2]
```diff
@@ -692,6 +696,7 @@ object FunctionRegistry {
     expression[Sequence]("sequence"),
     expression[ArrayRepeat]("array_repeat"),
     expression[ArrayRemove]("array_remove"),
+    expression[ArrayPrepend]("array_prepend"),
```
Comment (navinvishy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:699]: Relevant diff here.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala [hunk 3]
```diff
@@ -1840,6 +1939,47 @@ class CollectionExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper
     checkEvaluation(ArrayRepeat(Literal("hi"), Literal(null, IntegerType)), null)
   }
 
+  test("Array prepend") {
```
Comment (navinvishy) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:1942]: Relevant diff here.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 4]
```diff
@@ -4018,10 +4181,22 @@ object functions {
    * @group collection_funcs
    * @since 2.4.0
    */
+
   def array_remove(column: Column, element: Any): Column = withExpr {
     ArrayRemove(column.expr, lit(element).expr)
   }
 
+  /**
```
Comment (navinvishy) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4189]: Relevant diff here.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 5]
```diff
@@ -2645,78 +2257,103 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
...
         "inputType" -> "\"DECIMAL(2,1)\"",
-        "requiredType" -> "\"INT\""
-      ),
-      queryContext = Array(ExpectedContext("", "", 0, 19, "array_repeat(a, 1.0)"))
-    )
+        "requiredType" -> "\"INT\""),
+      queryContext = Array(ExpectedContext("", "", 0, 19, "array_repeat(a, 1.0)")))
   }
 
+  test("array prepend") {
```
Comment (navinvishy) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:2264]: Relevant diff here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 6]
```diff
@@ -119,21 +117,24 @@ case class Size(child: Expression, legacySizeOfNull: Boolean)
     val value = child.eval(input)
     if (value == null) {
       if (legacySizeOfNull) -1 else null
-    } else child.dataType match {
-      case _: ArrayType => value.asInstanceOf[ArrayData].numElements()
-      case _: MapType => value.asInstanceOf[MapData].numElements()
-      case other => throw QueryExecutionErrors.unsupportedOperandTypeForSizeFunctionError(other)
-    }
+    } else
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:120]: Can we remove unrelated style changes?

Comment (navinvishy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Yeah there were too many of those. I've removed them now.

**Comment by LuciferYang** [comment]
The jira number should be `[SPARK-41233](https://issues.apache.org/jira/browse/SPARK-41233)`

**Comment by LuciferYang** [comment]
@HyukjinKwon `ArrayAppend` and `ArrayPrepend` are special cases of `ArrayInsert`. Is it better to implement each function independently or consider them as a whole?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 7]
```diff
@@ -1399,6 +1399,119 @@ case class ArrayContains(left: Expression, right: Expression)
...
+      case (l, _) if !ArrayType.acceptsType(l) =>
+        DataTypeMismatch(
+          errorSubClass = "UNEXPECTED_INPUT_TYPE",
+          messageParameters = Map(
+            "paramIndex" -> "1",
+            "requiredType" -> toSQLType(ArrayType),
+            "inputSql" -> toSQLExpr(left),
+            "inputType" -> toSQLType(left.dataType)))
+      case (ArrayType(e1, _), e2) if e1.sameType(e2) =>
+        TypeUtils.checkForOrderingExpr(e2, prettyName)
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1491]: Why need  check `e2` can be ordered ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 8]
```diff
@@ -1399,6 +1399,119 @@ case class ArrayContains(left: Expression, right: Expression)
...
+
+  override def nullSafeEval(arr: Any, value: Any): Any = {
+    val numberOfElements = arr.asInstanceOf[ArrayData].numElements()
+    if (numberOfElements + 1 > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
+      throw QueryExecutionErrors.concatArraysWithElementsExceedLimitError(numberOfElements)
+    }
+    val newArray = new Array[Any](numberOfElements + 1)
+    newArray(0) = value
+    var pos = 1
+    arr
```
[the remark this answers is not in the document]
Comment (zhengruifeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1427]: or use `ArrayData.foreach`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 7 above)
Comment (navinvishy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1491]: There seems to be an ordering requirement for `ArrayData` since it has a `compare` and is backed by a `scala.collection.Seq`. But I'm ok with just returning a `TypeCheckResult.TypeCheckSuccess` here. I followed what `ArrayContains` did here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 9]
```diff
@@ -1399,6 +1399,119 @@ case class ArrayContains(left: Expression, right: Expression)
...
+      })
+  }
+
+  override def prettyName: String = "array_prepend"
+  override protected def withNewChildrenInternal(
+                                                  newLeft: Expression,
+                                                  newRight: Expression): ArrayPrepend =
+    copy(left = newLeft, right = newRight)
+  override def dataType: DataType = left.dataType
+  override def checkInputDataTypes(): TypeCheckResult = {
```
[the remark this answers is not in the document]
Comment (navinvishy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1476]: I did what `ArrayContains` does. Maybe we should consolidate this, since it makes sense for many of these to do the same thing? eg. `ArrayContains`, `ArrayRemove`, `ArrayPrepend`, `ArrayAppend` etc.

**Comment by navinvishy** [comment]
> The jira number should be [SPARK-41233](https://issues.apache.org/jira/browse/SPARK-41233)

Changed. Thanks!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala [hunk 10]
```diff
@@ -1840,6 +1840,47 @@ class CollectionExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper
...
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:1843]:
```scala
- test("Array prepend") {
+ test("SPARK-41233: ArrayPrepend") {
```

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 11]
```diff
@@ -2651,6 +2651,58 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
     )
   }
 
+  test("array prepend") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:2654]:
```scala
- test("array prepend") {
+ test("SPARK-41233: array prepend") {
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala (hunk 10 above)
Comment (LuciferYang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala]: This is not a bug fix, also need to add jira number?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala [hunk 12]
```diff
@@ -1840,6 +1840,47 @@ class CollectionExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper
...
+  test("SPARK-41233: ArrayPrepend") {
+    val a0 = Literal.create(Seq(1, 2, 3, 4), ArrayType(IntegerType))
+    val a1 = Literal.create(Seq("a", "b", "c"), ArrayType(StringType))
+    val a2 = Literal.create(Seq.empty[Integer], ArrayType(IntegerType))
+    val a3 = Literal.create(null, ArrayType(StringType))
+
+    checkEvaluation(ArrayPrepend(a0, Literal(0)), Seq(0, 1, 2, 3, 4))
+    checkEvaluation(ArrayPrepend(a1, Literal("a")), Seq("a", "a", "b", "c"))
+    checkEvaluation(ArrayPrepend(a2, Literal(1)), Seq(1))
+    checkEvaluation(ArrayPrepend(a2, Literal(null, IntegerType)), null)
```
Comment (zhengruifeng) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:1852]: sorry for the late reply, we have a update on the null handling, that is:
1, if the array is NULL, returns NULL;
2, if the array is NOT NULL, the element is NULL, prepend the NULL value into the array.

would you mind updating this behavior? thanks

**Comment by zhengruifeng** [comment]
also cc @beliefer

**Comment by LuciferYang** [comment]
@navinvishy cloud you resolve the conflict?

**Comment by beliefer** [comment]
Could we wait array_append merged then this PR could follow some convention and make a better abstract ?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala (hunk 12 above)
Comment (navinvishy) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala]: @zhengruifeng sorry about the delay here. I've updated this to reflect the behavior you mentioned. Fixed the tests to show that as well. Thanks!

**Comment by LuciferYang** [comment]
> Could we wait array_append merged then this PR could follow some convention and make a better abstract ? @zhengruifeng @LuciferYang @HyukjinKwon

If it is possible, I think we should wait

**Comment by navinvishy** [comment]
> > Could we wait array_append merged then this PR could follow some convention and make a better abstract ? @zhengruifeng @LuciferYang @HyukjinKwon
> 
> If it is possible, I think we should wait

Looks like array_append has already been merged? https://github.com/apache/spark/pull/38865

**Comment by navinvishy** [comment]
> > > Could we wait array_append merged then this PR could follow some convention and make a better abstract ? @zhengruifeng @LuciferYang @HyukjinKwon
> > 
> > 
> > If it is possible, I think we should wait
> 
> Looks like array_append has already been merged? #38865

Also, I'm happy to try and unify `ArrayPrepend` and `ArrayAppend` (and maybe some others?) if that is useful.

**Comment by LuciferYang** [comment]
I found `ArrayAppend ` already merged

**Comment by zhengruifeng** [comment]
@navinvishy mind fix the scala linter failure?

**Comment by navinvishy** [comment]
> @navinvishy mind fix the scala linter failure?

Sorry for the delay @zhengruifeng . Fixing now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (zhengruifeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1438]: would you minding make this `nullSafeEval` method and `doGenCode` method more consistent with `ArrayAppend`:

```
  @transient protected lazy val elementType: DataType =
    inputTypes.head.asInstanceOf[ArrayType].elementType

  override protected def nullSafeEval(arr: Any, elementData: Any): Any = {
    val arrayData = arr.asInstanceOf[ArrayData]
    val numberOfElements = arrayData.numElements() + 1
    if (numberOfElements > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
      throw QueryExecutionErrors.concatArraysWithElementsExceedLimitError(numberOfElements)
    }
    val finalData = new Array[Any](numberOfElements)
    finalData.update(0, elementData)
    arrayData.foreach(elementType, (i: Int, v: Any) => finalData.update(i + 1, v))
    new GenericArrayData(finalData)
  }
```

I think we can make a better abstract later

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 13]
```diff
@@ -1399,6 +1399,149 @@ case class ArrayContains(left: Expression, right: Expression)
...
+      val pos = ctx.freshName("pos")
+      val allocation = CodeGenerator.createArrayData(
+        newArray,
+        right.dataType,
+        newArraySize,
+        s" $prettyName failed.")
+      val assignment =
+        CodeGenerator.createArrayAssignment(newArray, right.dataType, arr, pos, i, false)
+      val newElemAssignment =
+        CodeGenerator.setArrayElement(newArray, right.dataType, pos, value, Some(rightGen.isNull))
```
Comment (zhengruifeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1464]: I think we should not use `right.dataType` in `assignment ` and `newElemAssignment `.

may use the
```
  @transient protected lazy val elementType: DataType =
    inputTypes.head.asInstanceOf[ArrayType].elementType
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 14]
```diff
@@ -1399,6 +1399,149 @@ case class ArrayContains(left: Expression, right: Expression)
...
+    }
+  }
+
+  override def prettyName: String = "array_prepend"
+
+  override protected def withNewChildrenInternal(
+      newLeft: Expression, newRight: Expression): ArrayPrepend =
+    copy(left = newLeft, right = newRight)
+
+  override def dataType: DataType = left.dataType
```
[the remark this answers is not in the document]
Comment (zhengruifeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1508]: see https://github.com/apache/spark/commit/718b6b7ed8277d5f6577367ab0d49f60f9777df7

[the remark this answers is not in the document]
Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1508]: I guess there is a test case missing to capture this nullability case?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1438]: +1 to be consistent with `ArrayAppend`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (zhengruifeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1508]: Yes, I think so

[the remark this answers is not in the document]
Comment (navinvishy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: @zhengruifeng I've made this change and added a test for null cases for both left and right params.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 13 above)
Comment (navinvishy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: @zhengruifeng thanks, I've made this change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (navinvishy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Thanks, I've made this consistent with `ArrayAppend`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 15]
```diff
@@ -1399,6 +1399,119 @@ case class ArrayContains(left: Expression, right: Expression)
...
+          messageParameters = Map(
+            "functionName" -> toSQLId(prettyName),
+            "dataType" -> toSQLType(ArrayType),
+            "leftType" -> toSQLType(left.dataType),
+            "rightType" -> toSQLType(right.dataType)))
+    }
+  }
+  override def inputTypes: Seq[AbstractDataType] = {
+    (left.dataType, right.dataType) match {
+      case (_, NullType) => Seq.empty
```
[the remark this answers is not in the document]
Comment (navinvishy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: This matches what is in `ArrayAppend` now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 16]
```diff
@@ -1399,6 +1399,145 @@ case class ArrayContains(left: Expression, right: Expression)
     copy(left = newLeft, right = newRight)
 }
 
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage =
```
Comment (zhengruifeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1404]: please also document the null handling like `ArrayApend`:

```
 Type of element should be similar to type of the elements of the array.
      Null element is also appended into the array. But if the array passed, is NULL
      output is NULL
```

examples
```
      > SELECT _FUNC_(array(1, 2, 3, null), null);
       [null,1,2,3,null]
      > SELECT _FUNC_(CAST(null as Array<Int>), 2);
       NULL
```

File: python/pyspark/sql/functions.py [hunk 17]
```diff
@@ -7619,6 +7619,36 @@ def get(col: "ColumnOrName", index: Union["ColumnOrName", int]) -> Column:
...
+    :class:`~pyspark.sql.Column`
+        an array excluding given value.
+
+    Examples
+    --------
+    >>> df = spark.createDataFrame([([2, 3, 4],), ([],)], ['data'])
+    >>> df.select(array_prepend(df.data, 1)).collect()
+    [Row(array_prepend(data, 1)=[1, 2, 3, 4]), Row(array_prepend(data, 1)=[1])]
+    """
+    return _invoke_function("array_prepend", _to_java_column(col), element)
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions.py:7649]:
```suggestion
    return _invoke_function_over_columns("array_prepend", col, lit(value))
```

File: python/pyspark/sql/functions.py [hunk 18]
```diff
@@ -7619,6 +7619,36 @@ def get(col: "ColumnOrName", index: Union["ColumnOrName", int]) -> Column:
     return _invoke_function_over_columns("get", col, index)
 
 
+@try_remote_functions
+def array_prepend(col: "ColumnOrName", element: Any) -> Column:
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions.py:7623]:
```suggestion
def array_prepend(col: "ColumnOrName", value: Any) -> Column:
```

to be consistent with `array_append`

File: python/pyspark/sql/functions.py [hunk 19]
```diff
@@ -7619,6 +7619,36 @@ def get(col: "ColumnOrName", index: Union["ColumnOrName", int]) -> Column:
...
+    at the beginning of the array.
+
+    .. versionadded:: 3.4.0
+
+    Parameters
+    ----------
+    col : :class:`~pyspark.sql.Column` or str
+        name of column containing array
+    element :
+        element to be prepended to the array
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions.py:7636]:
```suggestion
    value :
        a literal value, or a :class:`~pyspark.sql.Column` expression.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 20]
```diff
@@ -1399,6 +1399,145 @@ case class ArrayContains(left: Expression, right: Expression)
...
+    new GenericArrayData(finalData)
+  }
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val leftGen = left.genCode(ctx)
+    val rightGen = right.genCode(ctx)
+    val f = (arr: String, value: String) => {
+      val newArraySize = ctx.freshName("newArraySize")
+      val newArray = ctx.freshName("newArray")
+      val i = ctx.freshName("i")
+      val pos = ctx.freshName("pos")
```
Comment (zhengruifeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1451]: I guess we only need one of var `i` and `pos`?

Comment (navinvishy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: @zhengruifeng I assume you mean that we could use `i+1` here for the assignment instead of using a different var `pos`? I've done that here - please let me know if that works.

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
cc @dongjoon-hyun @beliefer @LuciferYang @cloud-fan @HyukjinKwon would you mind taking another look?

**Comment by navinvishy** [comment]
@dongjoon-hyun @beliefer @LuciferYang @cloud-fan @HyukjinKwon gentle ping to please take a look. Thanks!

**Comment by zhengruifeng** [comment]
@navinvishy  sorry, it seems late for 3.4.0.
would you mind changing the version from `3.4.0` to `3.5.0`?

I am going to merge this PR to master this week if no more comments.

**Comment by navinvishy** [comment]
> @navinvishy sorry, it seems late for 3.4.0. would you mind changing the version from `3.4.0` to `3.5.0`?
> 
> I am going to merge this PR to master this week if no more comments.

I've updated the version. Thanks @zhengruifeng !

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 21]
```diff
@@ -1399,6 +1399,151 @@ case class ArrayContains(left: Expression, right: Expression)
     copy(left = newLeft, right = newRight)
 }
 
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = """
+      _FUNC_(array, element) - Add the element at the beginning of the array passed as first
+      argument. Type of element should be similar to type of the elements of the array.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1406]:
```suggestion
      argument. Type of element should be the same to the type of the elements of the array.
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1406]: We don't need to document the type coercion behavior in each function. We have a dedicated doc to explain what "similar" type is: https://spark.apache.org/docs/latest/sql-ref-ansi-compliance.html#type-promotion-and-precedence

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 22]
```diff
@@ -1399,6 +1399,151 @@ case class ArrayContains(left: Expression, right: Expression)
...
+    val f = (arr: String, value: String) => {
+      val newArraySize = ctx.freshName("newArraySize")
+      val newArray = ctx.freshName("newArray")
+      val i = ctx.freshName("i")
+      val iPlus1 = s"$i+1"
+      val zero = "0"
+      val allocation = CodeGenerator.createArrayData(
+        newArray,
+        elementType,
+        newArraySize,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1464]: do we really need this variable? can we just pass `s"$arr.numElements() + 1"`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 23]
```diff
@@ -1399,6 +1399,151 @@ case class ArrayContains(left: Expression, right: Expression)
...
+           |${ev.isNull} = false;
+           |${resultCode}
+           |""".stripMargin
+      }
+      ev.copy(code =
+        code"""
+        boolean ${ev.isNull} = true;
+        ${CodeGenerator.javaType(dataType)} ${ev.value} = ${CodeGenerator.defaultValue(dataType)};
+        $nullSafeEval
+      """)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1493]: please please the same code style as
```
s"""
   |...
   |...
   |""". stripMargin
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 24]
```diff
@@ -1399,6 +1399,151 @@ case class ArrayContains(left: Expression, right: Expression)
...
+    } else {
+      ev.copy(code =
+        code"""
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1496]: ditto

**Review by cloud-fan — APPROVED** [review]

**Comment by zhengruifeng** [comment]
@navinvishy would you mind addressing wenchen's comments? we can merge it then.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 22 above)
Comment (navinvishy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1464]: Yes, that wasn't really necessary. I've removed it, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 23 above)
Comment (navinvishy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Changed this, thanks!

**Comment by navinvishy** [comment]
> @navinvishy would you mind addressing wenchen's comments? we can merge it then.

I've addressed them. Thanks for checking, @zhengruifeng !

**Comment by zhengruifeng** [comment]
merged into master, thank you @navinvishy for working on this!

**Comment by HyukjinKwon** [comment]
Seems like this broke the mima check for Spark connect. I am reverting this for now.

**Comment by LuciferYang** [comment]
@navinvishy like `ProblemFilters.exclude[Problem]("org.apache.spark.sql.functions.array_prepend"),` and you can run ` dev/connect-jvm-client-mima-check ` to check the result

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 25]
```diff
@@ -4043,6 +4043,16 @@ object functions {
   def array_compact(column: Column): Column = withExpr {
     ArrayCompact(column.expr)
   }
+    /**
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4046]: nit: The indentation of comments is different from others

Comment (navinvishy) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: @LuciferYang thanks. I've made the updates. There were some failures that seem to be unrelated to this change. I reran the workflow and seems to have succeeded this time.

Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: All Passed now :)

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Comment by zhengruifeng** [comment]
merged to master again 😄
